### ISSUE TITLE
Mobile Chrome: Cannot read property 'getInstance' of null

### ISSUE THREAD

**Issue by the-owl** [Pomax/react-onclickoutside#235]
This is very similar to #193.

This error appears on every click in the page after component has been unmounted.
Reproduced in Chrome 51 and below on Android. Can't be reproduced on desktop and in latest Chrome on Android versions (61).

Affected version: 6.6.0. Downgrade to 6.5.0 fixed this problem for me.
React version: 16
In my case this happens in webpack build without HMR.
Stack trace:
```
at onClickOutside.getInstance (vendor.js?hash=0488b695b563803f5fe3:39899:17)
at onClickOutside._this.__outsideClickHandler (vendor.js?hash=0488b695b563803f5fe3:39820:30)
at HTMLDocument.handlersMap.(anonymous function)
```

I have created a simple project that shows the issue: https://github.com/the-owl/react-onclickoutside-issue-repro
Run `npm start` & open the page in Chrome on Android <= 51, alert will appear on error.

**Comment by Andarist** [comment]
Thanks you for the provided testing project ❤️ 

Gonna check this out some time later.

**Comment by Andarist** [comment]
Any tips how I can debug this without having actual device? Im always using browserstack, but it seems that all mobile devices have Chrome 58 exclusively :s

**Comment by the-owl** [comment]
You can try UC Browser, which is present in BrowserStack. This issue can be reproduced on UC Browser too.

Replacing `WrappedComponent` with `TestComponent` (not wrapped by a HOC) on line 33 removes this error.
Tried `setInterval` with `console.log` - component unmounts correctly, it stops printing to console after clicking on button both on affected and unaffected browsers.
It always unmounts correctly, no matter what I do. I've tried to wrap `TestComponent` with a simple HOC that looks like this:
```
function simpleHOC (TestComponent) {
  return class extends Component {
    render () {
      return <TestComponent {...this.props} />;
    }
  };
}
```
Component mounts and unmounts correctly. Moving `componentDidMount`/`componentWillUnmount` in a HOC doesn't change anything.

Tried with React 15.5.0 - no effect, still reproducible.

**Comment by Andarist** [comment]
Thanks you for checking everything! Really - if only everyone would come from the start with repro repositories and would check everything maintainers suggest, the github would be better place :P 

Unfortunately I cant reproduce it on UC Browser on Browserstack - maybe it is some newer version too? Anyway - Im gonna probably just borrow some Android device from work tomorrow and debug it on it.

Thanks you for your report and help, it really is quite outstanding 👏

**Comment by dipnlik** [comment]
Just chiming in that I'm currently getting the same error in macOS Chrome in a typeahead field created by `react-bootstrap-typeahead`, which is using `react-onclickoutside` 6.6.0.  It is a private project so I can't simply share it but I can try to isolate the problem in a separate project if you want me to.

**Comment by Andarist** [comment]
macOS Chrome reproduction would be really great!

Although I literally just today got finally some Android device to test on, need to figure out how to install older Chrome on it. Probably we'll sit down to it today/tomorrow evening.

**Comment by dipnlik** [comment]
- https://dipnlik.github.io/typeahead-outside-235/
- https://github.com/dipnlik/typeahead-outside-235

**Comment by Andarist** [comment]
What are exact steps for the reproduction? Also could u specify exact versions of Chrome etc where this happens?

**Comment by dipnlik** [comment]
I'm using the latest Chrome version. Steps to reproduce are written directly on the linked page.

**Comment by Andarist** [comment]
🙈 I was in such a rush that I didnt take a proper look at those instructions. Thanks for the repro!

**Comment by Andarist** [comment]
@dipnlik I believe I have fixed your issue (already released and tested on your repro project)

@the-owl I dont think this was the same issue that you have experienced, but a confirmation from you would be great before I start working on your case.

**Comment by the-owl** [comment]
It seems that I've found the cause of the issue - it is related to passive listeners (d5e79fa1b9bd6e8685eb39663ea196568d668031). Older browsers do not understand options object as a third argument to `addEventListener`/`removeEventListener`. Old browsers expect `useCaption` as a third argument, and it is important to call `removeEventListener` with the same value of `useCaption` as in `addEventListener` call.

The value for third argument is an object that is passed only to `addEventListener`.
`removeEventListener` receives only two arguments. Furthermore, passing object of the same shape to `removeEventListener` does not work as expected - only passing _the same_ object works, like in this example (such code removes the issue, but it isn't a "clean" solution):
```
...
        // in enableOnClickOutside
	this._handlerOptions = {};
	
        events.forEach(eventName => {
          const isTouchEvent = touchEvents.indexOf(eventName) !== -1;

          if (isTouchEvent) {
            this._handlerOptions[eventName] = { passive: !this.props.preventDefault };
          }

          document.addEventListener(eventName, handlersMap[this._uid], this._handlerOptions[eventName]);
        });
...
        // in disableOnClickOutside
        return document.removeEventListener(eventName, fn, this._handlerOptions[eventName]);
...
```

I think that a proper fix for this issue will be to use [feature detection](https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md#feature-detection) and pass third argument to `addEventListener` only in browsers that support passive listeners.

**Comment by Andarist** [comment]
While fixing @dipnlik's case and looking at the source I thought it might be exactly issue like this! I've stumbled before on the same thing with PhantomJS.

I'd like to reuse this feature detection from somewhere (rather than inline it in the project) - I've created https://github.com/rafrex/detect-passive-events/pull/4 and https://github.com/rafrex/detect-passive-events/pull/5 . If they land I'm gonna make the other lib a dependency and gonna release a fix for this. If not I'm gonna inline it for the time being.
